File: Runtime/Common/LuaAppDomain.cs
// Copyright 2026 Code Philosophy
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

using System;
using System.Reflection;

namespace ZLua
{
    /// <summary>
    /// Host-facing facade. Concrete work is performed by an <see cref="ILuaRuntime"/>
    /// created on demand from <c>ZLua.Mono</c> (Editor) or <c>ZLua.Il2Cpp</c> (Player).
    /// </summary>
    public static class LuaAppDomain
    {
        private static ILuaRuntime s_runtime;
        private static Func<string, object> s_pendingResetLoader;

        public static void Initialize(Func<string, object> moduleLoader)
        {
            EnsureRuntime();
            s_runtime.Initialize(moduleLoader);
            LuaFramePump.EnsureRegistered();
        }

        /// <summary>
        /// Schedule a rebuild of the single main <c>lua_State</c>.
        /// Teardown / re-init runs at Unity EndOfFrame via <see cref="LuaFramePump"/> (not immediately).
        /// After it applies, prior <see cref="GetFunction{T}"/> delegates are invalid.
        /// </summary>
        public static void Reset(Func<string, object> moduleLoader)
        {
            if (moduleLoader == null)
            {
                throw new ArgumentNullException(nameof(moduleLoader));
            }

            EnsureRuntime();
            s_pendingResetLoader = moduleLoader;
            LuaFramePump.EnsureRegistered();
        }

        /// <summary>
        /// Bind a Lua module function to a closed delegate of type <typeparamref name="T"/>.
        /// Must be called after <see cref="Initialize"/>. Does not guarantee instance reuse across calls.
        /// </summary>
        public static T GetFunction<T>(string luaModule, string luaMethodName)
            where T : MulticastDelegate
        {
            EnsureRuntime();
            if (string.IsNullOrEmpty(luaModule))
            {
                throw new ArgumentException("luaModule must be non-empty.", nameof(luaModule));
            }

            if (string.IsNullOrEmpty(luaMethodName))
            {
                throw new ArgumentException("luaMethodName must be non-empty.", nameof(luaMethodName));
            }

            Delegate bound = s_runtime.GetFunction(typeof(T), luaModule, luaMethodName);
            return (T)bound;
        }

        internal static void ProcessPendingRefReleases()
        {
            s_runtime?.ProcessPendingRefReleases();
        }

        /// <summary>
        /// Apply a scheduled <see cref="Reset"/> if any. Called from <see cref="LuaFramePump"/> at EndOfFrame.
        /// </summary>
        internal static void FlushPendingReset()
        {
            if (s_pendingResetLoader == null || s_runtime == null)
            {
                return;
            }

            Func<string, object> loader = s_pendingResetLoader;
            s_pendingResetLoader = null;
            s_runtime.Reset(loader);
            LuaFramePump.EnsureRegistered();
        }

        private static void EnsureRuntime()
        {
            if (s_runtime != null)
            {
                return;
            }

#if UNITY_EDITOR
            const string hostTypeName = "ZLua.LuaMonoAppDomain, ZLua.Mono";
#else
            const string hostTypeName = "ZLua.LuaIl2CppAppDomain, ZLua.Il2Cpp";
#endif
            Type hostType = Type.GetType(hostTypeName);
            if (hostType == null)
            {
                throw new InvalidOperationException(
                    $"ZLua backend type not found: '{hostTypeName}'. Ensure the matching package assembly is loaded.");
            }

            Type runtimeType = hostType.GetNestedType("Runtime", BindingFlags.NonPublic);
            if (runtimeType == null || !typeof(ILuaRuntime).IsAssignableFrom(runtimeType))
            {
                throw new InvalidOperationException(
                    $"ZLua backend '{hostTypeName}' is missing a non-public nested Runtime : ILuaRuntime.");
            }

            s_runtime = (ILuaRuntime)Activator.CreateInstance(runtimeType, nonPublic: true);
        }
    }
}

File: ZLua~/zlua-runtime/lvm/LuaAppDomain.cpp
// Copyright 2026 Code Philosophy
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "LuaAppDomain.h"

#include "LuaEnv.h"
#include "LuaInternalCalls.h"
#include "LuaLoader.h"
#include "../utils/MetadataUtil.h"
#include "../utils/LuaMetadataAlloc.h"
#include "../utils/LuaException.h"
#include "../bridge/PropertyBridge.h"
#include "../bridge/MethodBridge.h"
#include "../bridge/DelegateBridge.h"
#include "../generated/MarshalBindings.h"
#include "../generated/CompositeSpecializedBindings.h"
#include "../generated/AliasBindings.h"
#include "../generated/ExtensionBindings.h"

#include "vm/Exception.h"

namespace zlua
{
static bool s_processInitialized = false;
static bool s_hostInitialized = false;

void LuaAppDomain::InitializeProcessOnce()
{
    if (s_processInitialized)
    {
        return;
    }

    LuaMetadataAlloc::Initialize();
    MetadataUtil::Initialize();
    PropertyBridge::Initialize();
    MethodBridge::Initialize();
    DelegateBridge::Initialize();
    RegisterMarshalBindingTables();
    RegisterCompositeSpecializedWriters();
    RegisterAliasBindingTables();
    RegisterExtensionBindingTables();
    LuaInternalCalls::RegisterCoreInternalCalls();
    LuaLoader::RegisterRoots();
    s_processInitialized = true;
}

void LuaAppDomain::InitializeState()
{
    IL2CPP_ASSERT(LuaEnv::GetState() == nullptr);
    LuaEnv::Initialize();
}

void LuaAppDomain::ShutdownState()
{
    if (LuaEnv::GetState() == nullptr)
    {
        s_hostInitialized = false;
        return;
    }

    LuaEnv::Shutdown();
    s_hostInitialized = false;
}

void LuaAppDomain::Initialize()
{
    InitializeProcessOnce();
    if (LuaEnv::GetState() == nullptr)
    {
        InitializeState();
    }
}

void LuaAppDomain::InitializeFromManaged(Il2CppDelegate* moduleLoaderDelegate)
{
    if (moduleLoaderDelegate == nullptr)
    {
        il2cpp::vm::Exception::Raise(il2cpp::vm::Exception::GetArgumentNullException("loader delegate is null"));
    }

    InitializeProcessOnce();
    if (LuaEnv::GetState() == nullptr)
    {
        InitializeState();
    }
    else if (s_hostInitialized)
    {
        il2cpp::vm::Exception::Raise(il2cpp::vm::Exception::GetInvalidOperationException(
            "ZLua is already initialized. Call LuaAppDomain.Reset to rebuild the Lua domain."));
    }

    LuaLoader::SetModuleLoader(moduleLoaderDelegate);
    s_hostInitialized = true;
}

void LuaAppDomain::ResetFromManaged(Il2CppDelegate* moduleLoaderDelegate)
{
    if (moduleLoaderDelegate == nullptr)
    {
        il2cpp::vm::Exception::Raise(il2cpp::vm::Exception::GetArgumentNullException("loader delegate is null"));
    }

    InitializeProcessOnce();
    ShutdownState();
    InitializeState();
    LuaLoader::SetModuleLoader(moduleLoaderDelegate);
    s_hostInitialized = true;
}
} // namespace zlua
